micro_mg3_0.F90

```diff
@@ -3088,6 +3143,161 @@ subroutine micro_mg_tend ( &
   end do
   !$acc end parallel
 
+! ----------------------------------------------
+!  Implicit sedimentation (AG, 7/2021: needs acc directives)      
+! ----------------------------------------------
+
+  do k=1,nlev
+     do i=1,mgncol
+       pdel_inv(i,k) = 1._r8/pdel(i,k)
+     enddo
+  enddo
+
+!     calculate midpoint height. Hypsometric equation
+!     first estimate midpoint height
+  do i=1,mgncol
+     ps = p(i,nlev) + pdel(i,nlev)/2._r8   !pseudo surface pressure   
+
+     zhalf(i,nlev)= r*t(i,nlev-1)/g*log(ps/p(i,nlev))      
+
+     do k = nlev-1,1,-1
+       H = r*t(i,k)/g*log(p(i,k+1)/p(i,k))
+       zhalf(i,k)=zhalf(i,k+1)+H
+     enddo
+  enddo
+
+! Implicit Sedimentation calculation: from Guo et al, 2021, GFDL version. 
+      
+if ( do_implicit_fall ) then
+      fc  = vfac_drop * fc/g/rho
+      fnc = vfac_drop * fnc/g/rho
+      fi  = vfac_ice  * fi/g/rho
+      fni = vfac_ice  * fni/g/rho
+
+      fr  = vfactor * fr/g/rho
+      fnr = vfactor * fnr/g/rho
+      fs  = vfactor * fs/g/rho
+      fns = vfactor * fns/g/rho
+      fg  = vfactor * fg/g/rho
+      fng = vfactor * fng/g/rho
+      
+ ! cloud water (mass) sedimentation
+      do i=1,mgncol
+         dum_1D(:) = dumc(i,:)
+         call implicit_fall ( deltat, 1, nlev, zhalf(i,:) , fc(i,:), pdel(i,:), dum_1D, precip, flx)
+         do k=1,nlev
+            lflx(i,k+1) = lflx(i,k+1) + flx(k)/g/deltat
+            qcsedten(i,k)= qcsedten(i,k) + (dum_1D(k) - dumc(i,k))/deltat
+            qctend(i,k)  = qctend(i,k)   + (dum_1D(k) - dumc(i,k))/deltat
+         enddo
+         if ( precip .ge. 0.0 ) then !h1g, 2019-11-26, ensure numerical stability
+            prect(i) = prect(i)+precip/g/deltat/1000._r8  
+         endif
+      enddo
```

Comment:
Could we push the loop of `mgncol` into the `implicit_fall` subroutine?

Reply:
Not sure. I'm going to leave this for now to check everything else. I followed the coding in another model. Maybe we could push it down, but leaving it for now.